os/mm/mm_heap/mm_partition_mgr.c

```diff
@@ -0,0 +1,159 @@
+/* If multiple regions are supported or if multi-heap is supported in the project,
+then we dont know how to configure the partitions for user applications. Hence we
+will throw an error here
+*/
+
+#if defined(CONFIG_APP_BINARY_SEPARATION) && defined(__KERNEL__)
+
+#if CONFIG_MM_REGIONS > 1
+#error "Unable to handle APP RAM PARTITIONS for CONFIG_MM_REGIONS > 1"
+#endif
+
+#if CONFIG_MM_NHEAPS > 1
+#error "Unable to handle APP RAM PARTITIONS for CONFIG_MM_NHEAPS > 1"
+#endif
+
+#define ALIGN_EVEN_MASK		2
+#define ALIGN_EVEN(a)		((a) + ALIGN_EVEN_MASK) & ~ALIGN_EVEN_MASK
+/****************************************************************************
+ * Public Variables
+ ****************************************************************************/
+extern uint32_t __ksram_segment_start__[];
+extern uint32_t __ksram_segment_size__[];
+
+/****************************************************************************
+ * Private variables
+ ****************************************************************************/
+static struct mm_partition_s g_mempartitions[CONFIG_NUM_APPS];
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+/****************************************************************************
+ * Name: mm_initialize_partitions
+ *
+ * Description:
+ * This function is called during os boot to initialize the memory partitions.
+ *
+ * Parameters:
+ *   None
+ *
+ * Return Value:
+ *   None
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+void mm_initialize_partitions(void)
+{
+	int i = 0;
+	uint32_t start;
+	uint32_t size;
+
+	start = (uint32_t)__ksram_segment_start__ + (uint32_t)__ksram_segment_size__;
+	size = (REGION_SIZE - start) / CONFIG_NUM_APPS;
+	for (i = 0; i < CONFIG_NUM_APPS; i++) {
+		g_mempartitions[i].start = ALIGN_EVEN(start);
+		g_mempartitions[i].size = size;
+		g_mempartitions[i].status = MM_PART_FREE;
+		start += size;
+	}
+}
+
+/****************************************************************************
+ * Name: mm_allocate_partition
+ *
+ * Description:
+ * This function is called to allocate a user app partition in which the app
+ * will be loaded by the loader.
+ *
+ * Parameters:
+ *   start_addr : OUT param to hold start address of allocated partition
+ *   size	: OUT param to hold size of allocated partition
+ *
+ * Return Value:
+ *   success: Zero value if partition was successfuly allocated.
+ *
+ * Assumptions:
+ *
+ ****************************************************************************/
+
+int8_t mm_allocate_partition(uint32_t **start_addr, uint32_t *size)
+{
+	int i;
+
+	for (i = 0; i < CONFIG_NUM_APPS; i++) {
```

Comment:
Is it okay not to be atomic?

Reply:
The atomicity can be handled in the loader section.